File: O(ElogV)_O(ElogE)/dijskstra/dijkstra.py
import heapq

def dijkstra(grafo, inicio):
    # Inicialização das distâncias com infinito
    distancias = {v: float('infinity') for v in grafo}
    distancias[inicio] = 0
    
    # Fila de Prioridade (Min-Heap)
    # Armazena tuplas: (distância_acumulada, vértice_atual)
    pq = [(0, inicio)]

    while pq:
        # --- PARTE DO log(V) ---
        # A operação heappop custa O(log V) ou O(log E).
        # Ela remove o vértice com a menor distância da fila.
        dist_atual, u = heapq.heappop(pq)

        # Se encontramos um caminho mais longo do que o já conhecido, ignoramos (Lazy Deletion)
        if dist_atual > distancias[u]:
            continue

        # --- PARTE DO E ---
        # Este loop percorre os vizinhos. Somando todas as iterações do while,
        # passaremos por todas as arestas (E) do grafo.
        for vizinho, peso in grafo[u].items():
            distancia = dist_atual + peso

            # Relaxamento da aresta
            if distancia < distancias[vizinho]:
                distancias[vizinho] = distancia
                
                # --- PARTE DO log(V) ---
                # A operação heappush insere na fila custando O(log V) ou O(log E).
                # Como isso pode acontecer para cada aresta, temos E * log V.
                heapq.heappush(pq, (distancia, vizinho))

    return distancias

'''
CÁLCULO DA COMPLEXIDADE (PARA O TCC):
1. Extrair Mínimo: O loop principal executa V vezes para extrair o mínimo.
   Custo: V * log V
2. Relaxar Arestas: O loop interno percorre cada aresta (E) do grafo.
   Para cada aresta, podemos fazer uma inserção no Heap (log V).
   Custo: E * log V
   
Total = (V * log V) + (E * log V)
Como geralmente E >= V - 1 em grafos conexos, simplificamos para:
RESULTADO = O(E log V)
'''

# Grafo (Lista de Adjacência)
grafo = {
    'A': {'B': 4, 'C': 2},
    'B': {'D': 3, 'E': 1},
    'C': {'B': 1, 'D': 5},
    'D': {'E': 2},
    'E': {}
}

print(dijkstra(grafo, 'A'))

File: O(ElogV)_O(ElogE)/dijskstra/Dijkstra.java
import java.util.*;

class Node implements Comparable<Node> {
    public int vertice;
    public int custo;

    public Node(int vertice, int custo) {
        this.vertice = vertice;
        this.custo = custo;
    }

    // Define que a prioridade é o menor custo (Min-Heap)
    @Override
    public int compareTo(Node other) {
        return Integer.compare(this.custo, other.custo);
    }
}

public class Dijkstra {
    // Método Dijkstra O(E log V)
    public static void calcular(List<List<Node>> adj, int src, int V) {
        // Heap de Prioridade (Min-Heap)
        PriorityQueue<Node> pq = new PriorityQueue<>();
        int[] dist = new int[V];
        Arrays.fill(dist, Integer.MAX_VALUE);

        pq.add(new Node(src, 0));
        dist[src] = 0;

        while (!pq.isEmpty()) {
            // --- PARTE DO log(V) ---
            // poll() remove o menor elemento em tempo logarítmico
            Node atual = pq.poll();
            int u = atual.vertice;

            // --- PARTE DO E ---
            // Itera sobre os vizinhos (arestas conectadas a u)
            for (Node vizinho : adj.get(u)) {
                int v = vizinho.vertice;
                int peso = vizinho.custo;

                if (dist[u] + peso < dist[v]) {
                    dist[v] = dist[u] + peso;
                    
                    // --- PARTE DO log(V) ---
                    // add() insere na fila de prioridade em tempo logarítmico
                    // No pior caso, fazemos isso para cada aresta (E)
                    pq.add(new Node(v, dist[v]));
                }
            }
        }

        // Impressão (apenas visualização)
        System.out.println("Distâncias a partir do vértice " + src + ":");
        for (int i = 0; i < V; i++) {
            System.out.println(i + " \t\t " + dist[i]);
        }
    }

    public static void main(String[] args) {
        int V = 5;
        List<List<Node>> adj = new ArrayList<>();
        for (int i = 0; i < V; i++) adj.add(new ArrayList<>());

        // Adicionando arestas (Exemplo: 0 -> 1 com peso 9)
        adj.get(0).add(new Node(1, 9));
        adj.get(0).add(new Node(2, 6));
        adj.get(0).add(new Node(3, 5));
        adj.get(0).add(new Node(4, 3));
        adj.get(2).add(new Node(1, 2));
        adj.get(2).add(new Node(3, 4));

        calcular(adj, 0, V);
    }
}